Python/ceval.c

```diff
@@ -781,27 +781,42 @@ _PyObjectArray_Free(PyObject **array, PyObject **scratch)
 /* This setting is reversed below following _PyEval_EvalFrameDefault */
 #endif
 
+#ifdef Py_TAIL_CALL_INTERP
+#include "generated_tail_call_handlers.c.h"
+static inline PyObject *_TAIL_CALL_shim(TAIL_CALL_PARAMS)
+{
+    opcode = next_instr->op.code;
+#ifdef LLTRACE
+    return (INSTRUCTION_TABLE[opcode])(frame, stack_pointer, tstate, next_instr, opcode, next_instr->op.arg, lltrace);
+#else
+    return (INSTRUCTION_TABLE[opcode])(frame, stack_pointer, tstate, next_instr, opcode, next_instr->op.arg);
+#endif
+}
+#endif
+
 PyObject* _Py_HOT_FUNCTION
 _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int throwflag)
 {
     _Py_EnsureTstateNotNULL(tstate);
     CALL_STAT_INC(pyeval_calls);
 
-#if USE_COMPUTED_GOTOS
+#if USE_COMPUTED_GOTOS && !defined(Py_TAIL_CALL_INTERP)
```

Comment:
The `USE_COMPUTED_GOTOS` and `Py_TAIL_CALL_INTERP` are mutually exclusive right? The `&& !defined(Py_TAIL_CALL_INTERP)` should then not be needed.

Reply:
No they're not! `USED_COMPUTED_GOTOS` detects whether labels as values is supported on a compiler. `Py_TAIL_CALL_INTERP` detects if efficient tail-calls are supported. So both can be on at the same time.

I think I'll guard and make sure it's either-or.

Reply:
On second thought. It's wrong to make it either-or. Some interpreters like the regex one uses computed gotos but not tail calls. So they can coexist.